### DISCUSSION TITLE
Guidelines for complex arguments

### DISCUSSION THREAD

**Discussion by jonhoo** [rust-lang-nursery/api-guidelines#187] [API Guidelines]
Builder are handy when you have a constructor many optional arguments (or arguments with defaults). However, in some cases you may have a constructor or method with many _required_ arguments. In those cases, we should probably recommend passing some kind of `Options` `struct` as an argument to the method with all the required parameters contained within it.

**Comment by dwijnand** [comment]
How do you construct the `Options` `struct`?

**Comment by jonhoo** [comment]
Not sure what you mean? I'm arguing that this:
```rust
Foo::new(Options {
  this: 800,
  that: true,
  other: 0.83,
  also: 20,
  btw: "do you even",
})
```
is probably better than
```rust
Foo::new(800, true, 0.83, 20, "do you even")
```
_especially_ when many of the arguments have the same type:
```rust
Foo::new(true, true, false, true)
```
And I think it'd be handy for the guidelines to give, well, guidelines, about how to deal with complex constructors like that.

**Comment by dwijnand** [comment]
I'm not sure exactly why, but I've seen people/advice (maybe even in these guidelines?) shying away from making struct fields public and allowing their construction like that.

Btw, there's a draft RFC that would help resolve this, IMO: https://github.com/Centril/rfcs/pull/19

**Comment by jonhoo** [comment]
I'm not sure exactly how that rfc helps? This is specifically for when the arguments are all _required_ (i.e., there is no default).

**Comment by dwijnand** [comment]
The idea is you'd set the default value for the non-required fields and would only define the fields you want to define when constructing the value.

**Comment by jonhoo** [comment]
Right, but that doesn't help when you have many required arguments :)

**Comment by dwijnand** [comment]
Why not?  (I'm sorry for making your issue here into a conversation with me 🙁)

IIUC it would make your example the following:

```rust
Foo {
  this: 800,
  that: true,
  other: 0.83,
  also: 20,
  btw: "do you even",
  optional_field: "I override your default value",
  launch_missiles: true,
}

**Comment by jonhoo** [comment]
Ohhhhhh, I see what you're saying. I'm thinking more generally of the idea of constructors. The very specific case of "all you want to do is set the initial field values" I agree that defaults solve. I'm more worried about when you have methods (they don't even have to be constructors) that take many arguments. Perhaps the use of `Foo::new` in my original post (and constructor in the title) was a mistake; I'll change that!

**Comment by dwijnand** [comment]
Ah right!  In that case it makes me think of @softprops' https://medium.com/@softprops/default-values-copy-that-ae43831781f3
